// @ciscospark/spark-core/src/interceptors/spark-tracking-id.js
/**!
 *
 * Copyright (c) 2015 Cisco Systems, Inc. See LICENSE file.
 */

import {Interceptor} from '@ciscospark/http-core';
import uuid from 'uuid';

export default class SparkTrackingIdInterceptor extends Interceptor {
  constructor(options) {
    // TODO set tracking id
    super();

    if (!options || !options.prefix) {
      throw new Error(`\`options.prefix\` is required. Did you forget to specify it at config.trackindIdPrefix?`);
    }

    let sequence = 0;
    const base = options.base || uuid.v4();
    const prefix = options.prefix || `spark-js-sdk`;

    Object.defineProperties(this, {
      base: {
        writable: false,
        value: base
      },
      prefix: {
        writable: false,
        value: prefix
      },
      sequence: {
        get: function get() {
          return sequence++;
        }
      }
    });
  }

  static create(options) {
    return new SparkTrackingIdInterceptor(options);
  }

  onRequest(options) {
    if (this.requiresTrackingId(options)) {
      options.headers.trackingid = this._generateTrackingId(options);
    }
    return options;
  }

  requiresTrackingId(options) {
    return !options.headers.trackingid;
  }

  _generateTrackingId() {
    return `${this.prefix}_${this.base}_${this.sequence}`;
  }
}
