// @ciscospark/http-core/dist/lib/interceptor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _promise = require("babel-runtime/core-js/promise");

var _promise2 = _interopRequireDefault(_promise);

var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require("babel-runtime/helpers/createClass");

var _createClass3 = _interopRequireDefault(_createClass2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**!
 *
 * Copyright (c) 2015 Cisco Systems, Inc. See LICENSE file.
 */

var Interceptor = function () {
  function Interceptor() {
    (0, _classCallCheck3.default)(this, Interceptor);
  }

  (0, _createClass3.default)(Interceptor, [{
    key: "onRequest",
    value: function onRequest(options) {
      return _promise2.default.resolve(options);
    }
  }, {
    key: "onRequestError",
    value: function onRequestError(options, reason) {
      return _promise2.default.reject(reason);
    }
  }, {
    key: "onResponse",
    value: function onResponse(options, response) {
      return _promise2.default.resolve(response);
    }
  }, {
    key: "onResponseError",
    value: function onResponseError(options, reason) {
      return _promise2.default.reject(reason);
    }
  }], [{
    key: "create",
    value: function create() {
      throw new Error("`Interceptor.create()` must be defined");
    }
  }]);
  return Interceptor;
}();

exports.default = Interceptor;
